### DISCUSSION TITLE
Running `make test`

### DISCUSSION THREAD

**Discussion by malm0d** [silius-rs/silius#308] [Q&A] [answered]
Hi there, I'm interested in contributing to Silius, but I'm pretty new to AA so there's a lot I'm not sure about. I had forked the repository and am currently just trying to run `make test`, but I keep getting failures in `simulation_tests` and `tracer_tests`:

```
running 36 tests
test simulation_tests::account_can_reference_its_own_allowance_on_other_contract_balance_database ... FAILED
test simulation_tests::accept_plain_request_memory ... FAILED
test simulation_tests::account_can_reference_its_own_allowance_on_other_contract_balance_memory ... FAILED
test simulation_tests::access_self_struct_data_database ... FAILED
test estimate_gas_tests::estimate_with_zero ... FAILED
test simulation_tests::access_self_struct_data_memory ... FAILED
test simulation_tests::account_succeeds_referecing_its_own_balance_database ... FAILED
test simulation_tests::accept_plain_request_database ... FAILED
test simulation_tests::fail_if_referencing_self_token_balance_after_wallet_creation_database ... FAILED
test simulation_tests::fail_if_create_too_many_database ... FAILED
test simulation_tests::fail_with_bad_opcode_in_ctr_database ... FAILED
test simulation_tests::account_succeeds_referecing_its_own_balance_memory ... FAILED
test simulation_tests::fail_if_create_too_many_memory ... FAILED
test simulation_tests::fail_referencing_self_token_database ... FAILED
test simulation_tests::fail_referencing_self_token_memory ... FAILED
test simulation_tests::fail_if_referencing_self_token_balance_after_wallet_creation_memory ... FAILED
test simulation_tests::fail_with_bad_opcode_in_paymaster_database ... FAILED
test simulation_tests::fail_with_bad_opcode_in_ctr_memory ... FAILED
test simulation_tests::fail_with_inner_oog_revert_memory ... FAILED
test simulation_tests::fail_with_bad_opcode_in_validation_database ... FAILED
test simulation_tests::fail_with_inner_oog_revert_database ... FAILED
test simulation_tests::fail_with_bad_opcode_in_validation_memory ... FAILED
test simulation_tests::fail_with_bad_opcode_in_paymaster_memory ... FAILED
test simulation_tests::fail_with_unstaked_paymaster_returning_context_database ... FAILED
test simulation_tests::succeed_with_inner_revert_memory ... FAILED
test simulation_tests::fail_with_validation_recursively_calls_handle_ops_database ... FAILED
test simulation_tests::reject_unkown_rule_database ... FAILED
test simulation_tests::fail_with_unstaked_paymaster_returning_context_memory ... FAILED
test simulation_tests::succeed_with_inner_revert_database ... FAILED
test simulation_tests::reject_unkown_rule_memory ... FAILED
test tracer_tests::count_opcode_depth_bigger_than_1 ... FAILED
test simulation_tests::fail_with_validation_recursively_calls_handle_ops_memory ... FAILED
test tracer_tests::should_report_direct_use_of_gas_opcode ... FAILED
test tracer_tests::trace_exec_self_call_itself ... FAILED
test tracer_tests::trace_exec_self_should_revert ... FAILED
test tracer_tests::should_ignore_gas_used_as_part_of_call ... FAILED
```

I have Geth installed and is currently running in dev mode:
`cargo run --release -- node --eth-client-address http://127.0.0.1:8545 --mnemonic-file ${HOME}/.silius/0xd76209f615f69d4dff3288e17bc342e24148694f --beneficiary 0xd76209f615f69d4dff3288e17bc342e24148694f --entry-points 0x5FF137D4b0FDCD49DcA30c7CF57E578a026d2789 --http --ws`

Additionally, when I run `make lint` I get an error:
```
bash-3.2$ cargo +nightly fmt --all -- --check
cargo clippy --all -- -D warnings -A clippy::derive_partial_eq_without_eq -D clippy::unwrap_used -D clippy::uninlined_format_args
warning: silius v0.5.0-alpha (/Users/malcolmtan/Projects/silius/bin/silius) ignoring invalid dependency `cargo-sort` which is missing a lib target
    Finished dev [unoptimized + debuginfo] target(s) in 1.19s
cargo sort --check --workspace --grouped
Checking silius...
Checking silius...
Checking bundler...
Checking contracts...
Checking grpc...
Checking mempool...
Checking p2p...
Checking primitives...
Checking rpc...
Checking simple-account...
Checking storage...
Checking user-operation...
Checking tests...
cargo +nightly udeps --workspace
warning: silius v0.5.0-alpha (/Users/malcolmtan/Projects/silius/bin/silius) ignoring invalid dependency `cargo-sort` which is missing a lib target
warning: silius v0.5.0-alpha (/Users/malcolmtan/Projects/silius/bin/silius) ignoring invalid dependency `cargo-sort` which is missing a lib target
    Checking time v0.3.34
   Compiling ring v0.16.20
   Compiling ring v0.17.8
   Compiling prettyplease v0.2.17
   Compiling ethers-core v2.0.14 (https://github.com/Vid201/ethers-rs?branch=feat/patch#c9de1525)
   Compiling num-integer v0.1.46
    Checking ipnet v2.9.0
   Compiling eyre v0.6.12
    Checking rustls-webpki v0.101.7
   Compiling num-bigint v0.4.4
error[E0282]: type annotations needed for `Box<_>`
  --> /Users/malcolmtan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/time-0.3.34/src/format_description/parse/mod.rs:83:9
   |
83 |     let items = format_items
   |         ^^^^^
...
86 |     Ok(items.into())
   |              ---- type must be known at this point
   |
help: consider giving `items` an explicit type, where the placeholders `_` are specified
   |
83 |     let items: Box<_> = format_items
   |              ++++++++

   Compiling libsecp256k1-core v0.3.0
    Checking sct v0.7.1
   Compiling rustls v0.21.10
    Checking semver v1.0.22
   Compiling num_cpus v1.16.0
For more information about this error, try `rustc --explain E0282`.
error: could not compile `time` (lib) due to 1 previous error
warning: build failed, waiting for other jobs to finish...
make: *** [lint] Error 101
```

I have not made any changes to the codebase (am just trying to run the tests) so I'm not sure what's causing `simulation_tests` and `tracer_tests` to fail. Could I be missing some configuration or additional step? (I was able to create a wallet for the bundler and run it)

Thank you!

**Comment by Vid201** [comment, marked as answer]
> Hi there, I'm interested in contributing to Silius, but I'm pretty new to AA so there's a lot I'm not sure about. I had forked the repository and am currently just trying to run `make test`, but I keep getting failures in `simulation_tests` and `tracer_tests`:
> 
> ```
> running 36 tests
> test simulation_tests::account_can_reference_its_own_allowance_on_other_contract_balance_database ... FAILED
> test simulation_tests::accept_plain_request_memory ... FAILED
> test simulation_tests::account_can_reference_its_own_allowance_on_other_contract_balance_memory ... FAILED
> test simulation_tests::access_self_struct_data_database ... FAILED
> test estimate_gas_tests::estimate_with_zero ... FAILED
> test simulation_tests::access_self_struct_data_memory ... FAILED
> test simulation_tests::account_succeeds_referecing_its_own_balance_database ... FAILED
> test simulation_tests::accept_plain_request_database ... FAILED
> test simulation_tests::fail_if_referencing_self_token_balance_after_wallet_creation_database ... FAILED
> test simulation_tests::fail_if_create_too_many_database ... FAILED
> test simulation_tests::fail_with_bad_opcode_in_ctr_database ... FAILED
> test simulation_tests::account_succeeds_referecing_its_own_balance_memory ... FAILED
> test simulation_tests::fail_if_create_too_many_memory ... FAILED
> test simulation_tests::fail_referencing_self_token_database ... FAILED
> test simulation_tests::fail_referencing_self_token_memory ... FAILED
> test simulation_tests::fail_if_referencing_self_token_balance_after_wallet_creation_memory ... FAILED
> test simulation_tests::fail_with_bad_opcode_in_paymaster_database ... FAILED
> test simulation_tests::fail_with_bad_opcode_in_ctr_memory ... FAILED
> test simulation_tests::fail_with_inner_oog_revert_memory ... FAILED
> test simulation_tests::fail_with_bad_opcode_in_validation_database ... FAILED
> test simulation_tests::fail_with_inner_oog_revert_database ... FAILED
> test simulation_tests::fail_with_bad_opcode_in_validation_memory ... FAILED
> test simulation_tests::fail_with_bad_opcode_in_paymaster_memory ... FAILED
> test simulation_tests::fail_with_unstaked_paymaster_returning_context_database ... FAILED
> test simulation_tests::succeed_with_inner_revert_memory ... FAILED
> test simulation_tests::fail_with_validation_recursively_calls_handle_ops_database ... FAILED
> test simulation_tests::reject_unkown_rule_database ... FAILED
> test simulation_tests::fail_with_unstaked_paymaster_returning_context_memory ... FAILED
> test simulation_tests::succeed_with_inner_revert_database ... FAILED
> test simulation_tests::reject_unkown_rule_memory ... FAILED
> test tracer_tests::count_opcode_depth_bigger_than_1 ... FAILED
> test simulation_tests::fail_with_validation_recursively_calls_handle_ops_memory ... FAILED
> test tracer_tests::should_report_direct_use_of_gas_opcode ... FAILED
> test tracer_tests::trace_exec_self_call_itself ... FAILED
> test tracer_tests::trace_exec_self_should_revert ... FAILED
> test tracer_tests::should_ignore_gas_used_as_part_of_call ... FAILED
> ```
> 
> I have Geth installed and is currently running in dev mode: `cargo run --release -- node --eth-client-address http://127.0.0.1:8545 --mnemonic-file ${HOME}/.silius/0xd76209f615f69d4dff3288e17bc342e24148694f --beneficiary 0xd76209f615f69d4dff3288e17bc342e24148694f --entry-points 0x5FF137D4b0FDCD49DcA30c7CF57E578a026d2789 --http --ws`
> 
> Additionally, when I run `make lint` I get an error:
> 
> ```
> bash-3.2$ cargo +nightly fmt --all -- --check
> cargo clippy --all -- -D warnings -A clippy::derive_partial_eq_without_eq -D clippy::unwrap_used -D clippy::uninlined_format_args
> warning: silius v0.5.0-alpha (/Users/malcolmtan/Projects/silius/bin/silius) ignoring invalid dependency `cargo-sort` which is missing a lib target
>     Finished dev [unoptimized + debuginfo] target(s) in 1.19s
> cargo sort --check --workspace --grouped
> Checking silius...
> Checking silius...
> Checking bundler...
> Checking contracts...
> Checking grpc...
> Checking mempool...
> Checking p2p...
> Checking primitives...
> Checking rpc...
> Checking simple-account...
> Checking storage...
> Checking user-operation...
> Checking tests...
> cargo +nightly udeps --workspace
> warning: silius v0.5.0-alpha (/Users/malcolmtan/Projects/silius/bin/silius) ignoring invalid dependency `cargo-sort` which is missing a lib target
> warning: silius v0.5.0-alpha (/Users/malcolmtan/Projects/silius/bin/silius) ignoring invalid dependency `cargo-sort` which is missing a lib target
>     Checking time v0.3.34
>    Compiling ring v0.16.20
>    Compiling ring v0.17.8
>    Compiling prettyplease v0.2.17
>    Compiling ethers-core v2.0.14 (https://github.com/Vid201/ethers-rs?branch=feat/patch#c9de1525)
>    Compiling num-integer v0.1.46
>     Checking ipnet v2.9.0
>    Compiling eyre v0.6.12
>     Checking rustls-webpki v0.101.7
>    Compiling num-bigint v0.4.4
> error[E0282]: type annotations needed for `Box<_>`
>   --> /Users/malcolmtan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/time-0.3.34/src/format_description/parse/mod.rs:83:9
>    |
> 83 |     let items = format_items
>    |         ^^^^^
> ...
> 86 |     Ok(items.into())
>    |              ---- type must be known at this point
>    |
> help: consider giving `items` an explicit type, where the placeholders `_` are specified
>    |
> 83 |     let items: Box<_> = format_items
>    |              ++++++++
> 
>    Compiling libsecp256k1-core v0.3.0
>     Checking sct v0.7.1
>    Compiling rustls v0.21.10
>     Checking semver v1.0.22
>    Compiling num_cpus v1.16.0
> For more information about this error, try `rustc --explain E0282`.
> error: could not compile `time` (lib) due to 1 previous error
> warning: build failed, waiting for other jobs to finish...
> make: *** [lint] Error 101
> ```
> 
> I have not made any changes to the codebase (am just trying to run the tests) so I'm not sure what's causing `simulation_tests` and `tracer_tests` to fail. Could I be missing some configuration or additional step? (I was able to create a wallet for the bundler and run it)
> 
> Thank you!

Hey @malm0d , which version of Geth are you using? Can you please try with version `1.12.0` ?

**Reply by malm0d** [reply]
@Vid201 Ah, I was running v`1.14.3`. I changed to v`1.12.0` and it works now. Sorry for the false alarm!!

**Reply by malm0d** [reply]
@Vid201 Do you know what could be the cause of the following error?

```
error: failed to run custom build command for `reth-mdbx-sys v0.12.1-0 (https://github.com/paradigmxyz/reth.git?rev=aa6f2cb0610fb4fa0926b42cfed7f8ff51e0db8a#aa6f2cb0)`

Caused by:
  process didn't exit successfully: `/Users/malcolmtan/Projects/silius/target/debug/build/reth-mdbx-sys-388f5cb7d005d9e7/build-script-build` (exit status: 101)
  --- stderr
  thread 'main' panicked at /Users/malcolmtan/.cargo/registry/src/index.crates.io-6f17d22bba15001f/bindgen-0.61.0/ir/context.rs:859:9:
  "MDBX_version_info_struct_(unnamed_at_/Users/malcolmtan/_cargo/git/checkouts/reth-36d3ea1d1152b20c/aa6f2cb/crates/storage/libmdbx-rs/mdbx-sys/libmdbx/mdbx_h_673_3)" is not a valid Ident
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

This happened when I tried to run `make test` on a different computer (or rather, whenever it tries to build). 
Previously (from the posts above) I was using a macbook with M1 chip and it was fine, but am using a macbook with M2 chip now. Followed the exact same steps as I did previously but not sure why its throwing this error now.

`rustup show`:
```
rustup show
Default host: aarch64-apple-darwin
rustup home:  /Users/malcolmtan/.rustup

installed toolchains
--------------------

stable-aarch64-apple-darwin (default)
nightly-aarch64-apple-darwin
1.76.0-aarch64-apple-darwin

active toolchain
----------------

1.76.0-aarch64-apple-darwin (overridden by '/Users/malcolmtan/Projects/silius/rust-toolchain.toml')
rustc 1.76.0 (07dca489a 2024-02-04)
```

**Reply by Vid201** [reply]
Yes, this is a known problem: https://github.com/silius-rs/silius/issues/188

The problem is with old Reth deps for database, which should be updated (there is no problem with the latest Reth commit). But we can't update the deps yet, because Reth made some changes and you cannot use the database for custom tables which are need in Silius. There is a PR open on Reth for that (https://github.com/paradigmxyz/reth/pull/7832), and once it's resolved we can continue to fix the problem here (update Reth dep or change the database).

**Reply by malm0d** [reply]
@Vid201 I'm a little puzzled as to how I was able to get it running on my macbook air (M1, 2020) but not on my other macbook air (M2) which I use to work on personal projects. On the M1 machine, I uninstalled Rust, started the whole setup for Silius again and had no problem

**Reply by Vid201** [reply]
> @Vid201 I'm a little puzzled as to how I was able to get it running on my macbook air (M1, 2020) but not on my other macbook air (M2) which I use to work on personal projects. On the M1 machine, I uninstalled Rust, started the whole setup for Silius again and had no problem

Hmm, interesting.

If you have any more time to dig into this and find anything, please let me know!
